### PR TITLE
remove all functionality deprecated in 0.20

### PR DESCRIPTION
**Pull request by davidhewitt**
This removes all functionality which was deprecated in PyO3 0.20.

Following our norms we could have removed these bits in 0.22, but things were quite busy my side so I just forgot to get around to this!

I'll open a follow-up PR now to also remove 0.21's deprecations.

### CODE REVIEW

File: src/instance.rs [hunk 1]
```diff
@@ -1329,11 +1329,8 @@ impl<T> Py<T> {
     /// # }
     /// ```
     #[inline]
-    pub fn clone_ref(&self, _py: Python<'_>) -> Py<T> {
-        unsafe {
-            ffi::Py_INCREF(self.as_ptr());
-            Self::from_non_null(self.0)
-        }
+    pub fn clone_ref(&self, py: Python<'_>) -> Py<T> {
+        unsafe { Py::from_borrowed_ptr(py, self.0.as_ptr()) }
```
Comment (LilyFoote) [inline src/instance.rs:1333]: This doesn't seem related to deprecations to me. Should it land separately?

**Review by LilyFoote — APPROVED** [review]
Looks good to me!

File: src/instance.rs (hunk 1 above)
Comment (davidhewitt) [reply src/instance.rs:1333]: Ah that's an unintended revert of #4313, will fix. Good catch

File: tests/ui/deprecations.stderr [hunk 2]
```diff
@@ -1,43 +1,69 @@
...
+  --> tests/ui/deprecations.rs:17:4
+   |
+17 | fn pyfunction_option_3(_i: u32, _any: Option<i32>, _foo: Option<String>) {}
+   |    ^^^^^^^^^^^^^^^^^^^
+
+warning: function `pyfunction_option_4` is never used
+  --> tests/ui/deprecations.rs:20:4
+   |
+20 | fn pyfunction_option_4(
+   |    ^^^^^^^^^^^^^^^^^^^
```
Comment (LilyFoote) [inline tests/ui/deprecations.stderr:69]: `tests/ui/deprecations.rs` is failing because these lines are not in the output.

Comment (davidhewitt) [reply tests/ui/deprecations.stderr:69]: Hmm I'll check what I've messed up here...
